#952 - Fixed preview for code blocks marked with ```

danff opened the pull request:
#952 - Fixed preview for code blocks.
- Updated MarkdownTxtmark to iterate through input text and add trailing spaces (for line breaks) only for fragments outside code blocks to avoid invalid br elements;
- Updated MarkdownTxtmark class with methods for inserting html pre elements where needed to display line breaks in code fragments;
- Added test method to MarkdownTxtmark for testing code blocks.

karato commented:
@danff Thanks, let me find someone who can review this pull request

karato commented:
@krzyk review it

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -58,14 +60,30 @@
+    /**
+     * Html &lt;pre&gt;&ltcode&gt; fragment.
+     */
+    private static final String PRE_CODE = "<pre><code>";
```
@danff this constant is used only in one method, please make it a variable there

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -58,14 +60,30 @@
+    /**
+     * Html &lt;/code&gt;&lt/end&gt; fragment.
+     */
+    private static final String END_CODE_PRE = "</code></pre>";
```
@danff this constant is used only in one method, please make it a variable there

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -88,11 +106,44 @@ private static String formatLinks(final String txt) {
+                    .append(fragment)
+                        .append(marker);
+                } else {
+                    builder.append(makeLineBreakMarkedCode(fragment, markers));
```
@danff do you really need recursion to solve this?

danff replied:
@krzyk The idea is to exclude code blocks separated with 3 chars sequence from adding extra spaces (with existing method `makeLineBreak()`). But the problem still exist also for blocks separated with 2 chars and one char. So decided to exclude them too. Either writing the same code (with copy/paste) in the `else` clause or another call to this method with second separator. A much elegant code would be without recursion but with functional interfaces we have in java 8, but of course, this is not under discussion.

krzyk commented on netbout-web/src/test/java/com/netbout/rest/MarkdownTxtmarkTest.java:
```diff
@@ -197,6 +197,27 @@ public void formatsTextFragmentsToHtml() throws Exception {
+                "<p><pre><code>code\nanother line of code</code></pre></p>",
+            },
+        };
+        for (final String[] pair : texts) {
```
@danff why do you need array and iteration to test only one case?

danff replied:
@krzyk Will add testing strings also for code spans, not only blocks.

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -110,4 +161,19 @@ private static String makeLineBreak(final String txt) {
+     */
+    private static String fixedCodeBlocks(final String txt) {
+        return txt
+            .replace("<code>`\r\n", MarkdownTxtmark.PRE_CODE)
```
@danff could you use a regexp that would match all 3 cases for pre code, and another one for the end code pre?

danff replied:
@krzyk Missed that. And using regexp will make unnecessary using of variables or constants for html strings.

krzyk commented:
@danff please see my initial comments

danff commented:
@krzyk Updated the `fixedCodeBlocks()` method to use regexp patterns (first try failed as \R was introduces only in java 8). Using end of line pattern as it is described in java 8 docs for the \R not supported in java 7. Updated (and renamed) test method with more test strings to test also the code spans and not only blocks.

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -110,4 +153,22 @@ private static String makeLineBreak(final String txt) {
+        final String eol = "\\u000D\\u000A|[\\u000A\\u000B\\u000C\\u000D\\u0085\\u2028\\u2029]";
+        return txt
+            .replaceAll(
+                String.format("<code>`(%s)?", eol),
```
@danff and now there is a need to bring back static variables :) `Pattern.compile` the regex so it will be done only once, not per call to this method

krzyk commented on netbout-web/src/main/java/com/netbout/rest/MarkdownTxtmark.java:
```diff
@@ -110,4 +153,22 @@ private static String makeLineBreak(final String txt) {
+                String.format("<code>`(%s)?", eol),
+                "<pre><code>"
+            ).replaceAll(
+                String.format("(%s)?</code>`", eol),
```
@danff and here also, Pattern.compile into a static variable

krzyk commented:
@danff one (or two) comment regarding regex

danff commented:
@krzyk Made patterns compiled as static constants.

krzyk commented:
@danff thanks

krzyk commented:
@rultor merge pls

rultor commented:
> @rultor merge pls

@krzyk Thanks for your request. @dmzaytsev Please confirm this.

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/7547-193119713)

rultor commented:
> @rultor merge

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7547-193119713) (took me 8min)

dmzaytsev commented:
@rultor deploy

rultor commented:
> @rultor deploy

@dmzaytsev OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7547-193133853)

rultor commented:
> @rultor deploy

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7547-193133853) (took me 15min)

karato commented:
@ypshenychka please, review this ticket for compliance with our [QA rules](http://at.teamed.io/qa.html)

karato commented:
@rultor deploy pls

rultor commented:
> @rultor deploy pls

@karato OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7547-193778514)

rultor commented:
> @rultor deploy pls

@karato Done! FYI, the full log is [here](http://www.rultor.com/t/7547-193778514) (took me 12min)

ypshenychka commented:
@karato Quality is good.

karato commented:
> @karato Quality is good.

@ypshenychka thanks for the review

karato commented:
@krzyk since quality is good, 10 mins paid to @dmzaytsev (our architect), transaction `79940390`

karato commented:
@krzyk added 10 mins to @ypshenychka, for QA review, payment code is `79949017`... Thanks for your contribution, **26 mins** was added to your account, payment ID is `79949037`, 72 hours and 32 mins spent total... extra minutes for review comments (c=11)... +26 added to your rating, current score is: [+6106](http://www.netbout.com/b/35035?open=rating)
